### PR TITLE
Feature/#220 language menu

### PR DESCRIPTION
**Pull request by BranescuMihai**
---

### Pull Request checklist

- [X] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [X] **Screenshots**: This PR includes link to video at the bottom of this comment
- [X] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- [ ] **Milestone**: Make sure issues finished by this pull request are added to the [milestone](https://github.com/mozilla-mobile/fenix/milestones) of the version currently in development.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

[**Link with feature**](https://drive.google.com/open?id=1suU1pUNDwIuaSdoub2UpBVCzksyFstRs)

### CODE REVIEW

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    localeList.add(Locale("ta"))
+    localeList.add(Locale("tr"))
+    return localeList
+}
+
+/**
+ * Returns the locale that corresponds to the language stored locally by us. If no suitable one is found,
+ * return default.
+ */
+fun LocaleManager.getSelectedLocale(context: Context, localeList: List<Locale>): Locale {
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: `localeList` definitely makes this easier to test, but it doesn't seem to me that consumers should have to pass that in order to get the selected locale.  What about making it an optional param?  `localeList: List<Locale> = getSupportedLocales()`

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    val selectedLocale = getCurrentLocale(context)
+    if (selectedLocale == null) {
+        return localeList[0]
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Based on the kdoc, it sounds like this is assuming that the first entry will always be the default value.  Could we just return `getSystemDefault()` here instead?  Seems easier to understand to me.

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    } else {
+        for (locale in localeList) {
+            if (locale.toLanguageTag() == selectedLocale.toLanguageTag()) {
+                return locale
+            }
+        }
+    }
+    return localeList[0]
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Optional nit: this one's up to personal preference, but since `toLanguageTag` is guaranteed to be nonnull this method could be rewritten as

```kotlin
fun LocaleManager.getSelectedLocale(context: Context, localeList: List<Locale>): Locale {
    val selectedLocale = getCurrentLocale(context)?.toLanguageTag()
    val supportedMatch = localeList
        .firstOrNull { it.toLanguageTag() == selectedLocale }

    return supportedMatch ?: getSystemDefault()
}
```

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        super.onViewCreated(view, savedInstanceState)
+        consumeFrom(store) {
+            localeView.update(it)
+        }
+    }
+
+    private fun getStore(): LocaleSettingsStore {
+        val supportedLocales = LocaleManager.getSupportedLocales()
+        val selectedLocale: Locale
+        selectedLocale = LocaleManager.getSelectedLocale(requireContext(), supportedLocales)
```
Comment (Baron-Severin) [inline app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt]: Nit: `selectedLocale` assignment could be inlined with its declaration.

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    localeList.add(Locale("ru"))
+    localeList.add(Locale("sk"))
+    localeList.add(Locale("es"))
+    localeList.add(Locale("es", "AR"))
+    localeList.add(Locale("es", "ES"))
+    localeList.add(Locale("sv", "SE"))
+    localeList.add(Locale("ta"))
+    localeList.add(Locale("tr"))
+    return localeList
+}
```
[the remark this answers is not in the document]
Comment (Pike) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Is there a way to do this w/out hard-coding the list?

For one, this list is already out of date, we just added more locales.

Also, we might skip locales as part of the release builds, and then the UI would claim support for locales we don't ship, right?

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    private val interactor: LocaleSettingsViewInteractor,
+    private val selectedLocale: Locale
+) : BaseLocaleViewHolder(view) {
+    private val icon = view.locale_selected_icon
+    private val title = view.locale_title_text
+    private val subtitle = view.locale_subtitle_text
+
+    override fun bind(locale: Locale) {
+        title.text = locale.getDisplayName(locale).capitalize(locale)
+        subtitle.text = locale.displayName.capitalize(Locale.getDefault())
```
[the remark this answers is not in the document]
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt]: Yes, the reason is it is capitalising the text using the appropriate locale (for example, in Turkish, capital 'i' is not I). Will leave a comment:)

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt (hunk 4 above)
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleSettingsFragment.kt]: true, inlined

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+            icon.visibility = View.INVISIBLE
+        }
+
+        itemView.setOnClickListener {
+            interactor.onLocaleSelected(locale)
+        }
+    }
+}
+
+class DefaultLocaleViewHolder(
```
[the remark this answers is not in the document]
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt]: Changed to `SystemLocaleViewHolder`

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: In Focus and FFTV, we have a Gradle task that auto-generates this list based on the the different variants of `/res/values-{android_locale}/strings.xml` directories. We could do something similar here but ideally we were looking to create a Gradle plugin on Android Components to share this functionality between multiple apps, you can find more info here https://github.com/mozilla-mobile/android-components/issues/3586

[the remark this answers is not in the document]
Comment (Amejia481) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: This also another related https://github.com/mozilla-mobile/android-components/issues/3585 issue, hope it helps!

[the remark this answers is not in the document]
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: @Pike totally agree with removing the hardcoded list, it's just a shame Android has a method in `AssetManager.java` called `getNonSystemLocales()` that is exactly what we need, but it's in their internal API.
I had no idea we had a similar implementation on Focus and FFTV, so I created my own Gradle task, but I feel it's actually a bit simpler than those (for example, using build config instead of generating a java class). 
@Amejia481 please have a look and tell me your opinion :)

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+    private val subtitle = view.locale_subtitle_text
+
+    override fun bind(locale: Locale) {
+        title.text = locale.getDisplayName(locale).capitalize(locale)
+        subtitle.text = locale.displayName.capitalize(Locale.getDefault())
+
+        if (locale === selectedLocale) {
+            icon.visibility = View.VISIBLE
+        } else {
+            icon.visibility = View.INVISIBLE
```
[the remark this answers is not in the document]
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleAdapter.kt]: Changed, had to modify the goneMargin, but you're right, it's a bit cleaner this way, thanks!

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt (hunk 1 above)
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Yep, sounds right, they shouldn't. Added as optional

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt (hunk 2 above)
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Fair enough, changed to getSystemDefault instead of the first entry

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt (hunk 3 above)
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: Changed it, thanks!

File: app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (Pike) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: I left a note on the a-c issue, I came across https://developer.android.com/reference/android/content/res/Configuration.html#getLocales(). But you'd need to practically test it to see if it does the trick. Also in the case of us restricting resources we ship.

[the remark this answers is not in the document]
Comment (BranescuMihai) [reply app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt]: But regardless of which Gradle task we will use in the end, I agree, we should move it to AC so all 3 products can use it

File: app/build.gradle [hunk 9]
```diff
@@ -614,6 +615,23 @@ task printVariants {
     }
 }
 
+task buildTranslationArray {
```
Comment (Amejia481) [inline app/build.gradle]: Would you mind to move this to AC? :) We can have it as a Gradle plugin on the  `mozilla.components.support.locale`package, we have some examples of Gradle plugins in [buildSrc](https://github.com/psymoon/android-components/tree/08661a2af1d0afa09902daeec2423e40014006ce/buildSrc) and [here](https://github.com/psymoon/android-components/blob/c58d6a32d3c3815ced44fa65a811a42a51f0daee/build.gradle#L278). It will be pretty nice if we could generate an extension function on `LocaleManager` for people that use the plugging. Maybe [Kotlin Poet ](https://github.com/square/kotlinpoet#extension-functions)could be very helpful. What do you think?

Something like:

```kotlin
fun LocaleManager.getSupportedLocales(): List<Locale>{
    return listOf<Locale>(
            Locale("eu"),
            Locale("ca"),
            Locale("zh", "CN"),
            Locale("zh", "TW"),
            Locale("cs"),
            Locale("nl"),
            Locale("en"),
            Locale("fi"),
            Locale("fr"),
            Locale("de"),
            Locale("hu"),
            Locale("in"),
            Locale("it"),
            Locale("ja"),
            Locale("ko"),
            Locale("pl"),
            Locale("pa", "IN"),
            Locale("ru"),
            Locale("sk"),
            Locale("es"),
            Locale("es", "AR"),
            Locale("es", "ES"),
            Locale("sv", "SE"),
            Locale("ta"),
            Locale("tr")
    )
} 
```

Comment (BranescuMihai) [reply app/build.gradle]: Hmm, turns out that it's not as straight-forward as I hoped at all. Looked into Kotlin Poet and the class generation is the doable part, but the following issues appeared:
- From what I see, declaring the task as a plugin in buildSrc makes it harder for me to share across projects
- Fenix needs to have it's own task, that simply calls the locale task in AC - currently I hardcode an absolute path to the locale task in AC, but I don't know exactly how to do that with the actual AC dependency
- AC task needs to access the res folder in fenix(and eventually other projects) in order to get the locales, and for that we need a path - currently I can't find how to send the path as a param (maybe how we do with ktlint?) - [here is their implementation](https://github.com/pinterest/ktlint/blob/master/ktlint/src/main/kotlin/com/pinterest/ktlint/Main.kt)

**TLDR**: I need to call an AC task from Fenix with params. If you have experience with this, I would appreciate the help! Otherwise, I would suggest going forward with this PR as is, since it will be working for Fenix and we're kinda on a short schedule for the planned release

Comment (Amejia481) [reply app/build.gradle]: Let's me try to take a look at what you have done so far (Could you share your repo?). If it gets too complicated we can proceed as it is and improve it in a follow issue.

Comment (Amejia481) [reply app/build.gradle]: Please, proceed with what you have so far here. I will try to find an alternative on AC.

**Comment by Baron-Severin** [comment]
@BranescuMihai when you have a moment, would you mind resolving the conflicts and then landing this?

**Comment by Baron-Severin** [comment]
I think @BranescuMihai may be on PTO, but these conflicts are very minor.  I'll fix them up and merge this in.

**Comment by BranescuMihai** [comment]
> I think @BranescuMihai may be on PTO, but these conflicts are very minor. I'll fix them up and merge this in.

Yep, I was on PTO, thanks for helping with the conflicts @Baron-Severin !
